executor/cook/executor.py

```diff
@@ -0,0 +1,384 @@
+    task_id = get_task_id(task)
+    logging.debug('Updating task {} state to {}'.format(task_id, task_state))
+    status = create_status(task_id, task_state)
+    driver.sendStatusUpdate(status)
+
+
+def send_message(driver, message, max_message_length):
+    """
+        :param driver: The driver to send the message to.
+        :param message: The raw message to send.
+        :param max_message_length: The allowed max message length after encoding
+        :return: whether the message was successfully sent
+    """
+    logging.debug('Sending framework message {}'.format(message))
+    encoded_message = encode_data(str(message).encode('utf8'))
+    if len(encoded_message) < max_message_length:
+        driver.sendFrameworkMessage(encoded_message)
+        return True
+    else:
+        log_message_template = 'Unable to send message {} as it exceeds allowed max length of {}'
+        logging.info(log_message_template.format(message, max_message_length))
+        return False
+
+
+def launch_task(driver, task, stdout_name, stderr_name):
+    """
+        :param driver: The mesos driver to use.
+        :param task: The task to execute.
+        :param stdout_name: The file to use to redirect stdout.
+        :param stderr_name: The file to use to redirect stderr.
+        :return: The tuple containing the process launched, stdout file, and stderr file.
+    """
+    try:
+        update_status(driver, task, 'TASK_RUNNING')
+
+        command = decode_data(task['data']).decode('utf8')
+        logging.info('Command: {}'.format(command))
+        stdout = open(stdout_name, 'a+')
+        stderr = open(stderr_name, 'a+')
+        process = subprocess.Popen(command, shell=True, stdout=stdout, stderr=stderr)
+
+        return process, stdout, stderr
+    except Exception:
+        logging.exception('Error in launch_task')
+        update_status(driver, task, 'TASK_FAILED')
+
+
+def is_running(process):
+    """
+        :param process: The process to query
+        :return: whether the process is still running.
+    """
+    return process.poll() is None
+
+
+def cleanup_process(process_info):
+    """
+        :param process_info: Tuple of (process, stdout_file, stderr_file)
+        :return: Nothing
+    """
+    _, stdout, stderr = process_info
+    logging.info('Closing {}'.format(stdout.name))
+    stdout.close()
+    logging.info('Closing {}'.format(stderr.name))
+    stderr.close()
+
+
+def kill_task(driver, task, process_info):
+    """
+        Attempts to kill a process by sending it a SIGTERM.
+        If the process does not terminate inside a second, it is then sent a SIGKILL.
+        :param driver: The mesos driver to use.
+        :param task: The task to execute.
+        :param process_info: Tuple of (process, stdout_file, stderr_file)
+        :return: Nothing
+    """
+    try:
+        process, _, _ = process_info
+        if is_running(process):
+            logging.info('Waiting up to one second for process to terminate')
```

Comment:
Mesos uses `MESOS_EXECUTOR_SHUTDOWN_GRACE_PERIOD` to forcefully kill a task. I will change the code to use this value before force killing.